File: core/data/tests/externally_tagged_enum_renaming/output.kt
package com.agilebits.onepassword.colorsModule

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import kotlinx.serialization.descriptors.*
import kotlinx.serialization.encoding.*

/** Tests renaming variants */
@Serializable(CamelCaseEnumSerializer::class)
sealed class CamelCaseEnum {
	@Serializable
	object UnitVariant : CamelCaseEnum()
	
	@Serializable
	data class SimpleValue(val value: String) : CamelCaseEnum()
	
	@Serializable
	data class ComplexValue(val field: Int) : CamelCaseEnum()
}

class CamelCaseEnumSerializer : JsonContentPolymorphicSerializer<CamelCaseEnum>(CamelCaseEnum::class) {
	override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out CamelCaseEnum> {
		return when {
			element is JsonPrimitive && element.isString && element.content == "unitVariant" -> CamelCaseEnum.UnitVariant.serializer()
			element is JsonObject && "simpleValue" in element -> CamelCaseEnum.SimpleValue.serializer()
			element is JsonObject && "complexValue" in element -> CamelCaseEnum.ComplexValue.serializer()
			else -> throw SerializationException("Unknown element: $element")
		}
	}
}

/** Tests specific renames on variants */
@Serializable(SpecificRenameEnumSerializer::class)
sealed class SpecificRenameEnum {
	@Serializable
	object Regular : SpecificRenameEnum()
	
	@Serializable
	data class CustomName(val value: String) : SpecificRenameEnum()
	
	@Serializable
	data class CustomStruct(val value: Int) : SpecificRenameEnum()
}

class SpecificRenameEnumSerializer : JsonContentPolymorphicSerializer<SpecificRenameEnum>(SpecificRenameEnum::class) {
	override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out SpecificRenameEnum> {
		return when {
			element is JsonPrimitive && element.isString && element.content == "Regular" -> SpecificRenameEnum.Regular.serializer()
			element is JsonObject && "custom_name" in element -> SpecificRenameEnum.CustomName.serializer()
			element is JsonObject && "custom_struct" in element -> SpecificRenameEnum.CustomStruct.serializer()
			else -> throw SerializationException("Unknown element: $element")
		}
	}
}

/** Mixed kebab and specific renames */
@Serializable(MixedRenameEnumSerializer::class)
sealed class MixedRenameEnum {
	@Serializable
	object UnitValue : MixedRenameEnum()
	
	@Serializable
	data class CustomValue(val value: String) : MixedRenameEnum()
	
	@Serializable
	data class ComplexValue(val field: Int) : MixedRenameEnum()
}

class MixedRenameEnumSerializer : JsonContentPolymorphicSerializer<MixedRenameEnum>(MixedRenameEnum::class) {
	override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out MixedRenameEnum> {
		return when {
			element is JsonPrimitive && element.isString && element.content == "unit-value" -> MixedRenameEnum.UnitValue.serializer()
			element is JsonObject && "CUSTOM" in element -> MixedRenameEnum.CustomValue.serializer()
			element is JsonObject && "complex-value" in element -> MixedRenameEnum.ComplexValue.serializer()
			else -> throw SerializationException("Unknown element: $element")
		}
	}
}

File: core/data/tests/externally_tagged_enum_renaming/output.scala
package com.agilebits.onepassword

import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._

/** Tests renaming variants */
sealed trait CamelCaseEnum

object CamelCaseEnum {
	case object UnitVariant extends CamelCaseEnum
	case class SimpleValue(value: String) extends CamelCaseEnum
	case class ComplexValue(field: Int) extends CamelCaseEnum
	
	implicit val encoder: Encoder[CamelCaseEnum] = Encoder.instance {
		case UnitVariant => Json.fromString("unitVariant")
		case SimpleValue(value) => Json.obj("simpleValue" -> value.asJson)
		case ComplexValue(field) => Json.obj("complexValue" -> Json.obj("field" -> field.asJson))
	}
	
	implicit val decoder: Decoder[CamelCaseEnum] = new Decoder[CamelCaseEnum] {
		def apply(c: HCursor): Decoder.Result[CamelCaseEnum] = {
			if (c.value.isString && c.value.asString.contains("unitVariant")) {
				Right(UnitVariant)
			} else {
				val keys = c.keys.getOrElse(Nil)
				if (keys.isEmpty) {
					Left(DecodingFailure("Expected object with a single key", c.history))
				} else {
					val key = keys.head
					key match {
						case "simpleValue" => c.downField("simpleValue").as[String].map(SimpleValue)
						case "complexValue" => c.downField("complexValue").downField("field").as[Int].map(ComplexValue)
						case _ => Left(DecodingFailure(s"Unknown key: $key", c.history))
					}
				}
			}
		}
	}
}

/** Tests specific renames on variants */
sealed trait SpecificRenameEnum

object SpecificRenameEnum {
	case object Regular extends SpecificRenameEnum
	case class CustomName(value: String) extends SpecificRenameEnum
	case class CustomStruct(value: Int) extends SpecificRenameEnum
	
	implicit val encoder: Encoder[SpecificRenameEnum] = Encoder.instance {
		case Regular => Json.fromString("Regular")
		case CustomName(value) => Json.obj("custom_name" -> value.asJson)
		case CustomStruct(value) => Json.obj("custom_struct" -> Json.obj("value" -> value.asJson))
	}
	
	implicit val decoder: Decoder[SpecificRenameEnum] = new Decoder[SpecificRenameEnum] {
		def apply(c: HCursor): Decoder.Result[SpecificRenameEnum] = {
			if (c.value.isString && c.value.asString.contains("Regular")) {
				Right(Regular)
			} else {
				val keys = c.keys.getOrElse(Nil)
				if (keys.isEmpty) {
					Left(DecodingFailure("Expected object with a single key", c.history))
				} else {
					val key = keys.head
					key match {
						case "custom_name" => c.downField("custom_name").as[String].map(CustomName)
						case "custom_struct" => c.downField("custom_struct").downField("value").as[Int].map(CustomStruct)
						case _ => Left(DecodingFailure(s"Unknown key: $key", c.history))
					}
				}
			}
		}
	}
}

/** Mixed kebab and specific renames */
sealed trait MixedRenameEnum

object MixedRenameEnum {
	case object UnitValue extends MixedRenameEnum
	case class CustomValue(value: String) extends MixedRenameEnum
	case class ComplexValue(field: Int) extends MixedRenameEnum
	
	implicit val encoder: Encoder[MixedRenameEnum] = Encoder.instance {
		case UnitValue => Json.fromString("unit-value")
		case CustomValue(value) => Json.obj("CUSTOM" -> value.asJson)
		case ComplexValue(field) => Json.obj("complex-value" -> Json.obj("field" -> field.asJson))
	}
	
	implicit val decoder: Decoder[MixedRenameEnum] = new Decoder[MixedRenameEnum] {
		def apply(c: HCursor): Decoder.Result[MixedRenameEnum] = {
			if (c.value.isString && c.value.asString.contains("unit-value")) {
				Right(UnitValue)
			} else {
				val keys = c.keys.getOrElse(Nil)
				if (keys.isEmpty) {
					Left(DecodingFailure("Expected object with a single key", c.history))
				} else {
					val key = keys.head
					key match {
						case "CUSTOM" => c.downField("CUSTOM").as[String].map(CustomValue)
						case "complex-value" => c.downField("complex-value").downField("field").as[Int].map(ComplexValue)
						case _ => Left(DecodingFailure(s"Unknown key: $key", c.history))
					}
				}
			}
		}
	}
}